Cursor screenline public

chrisbra opened the pull request:
As requested in https://github.com/vim/vim/pull/4693#issuecomment-529642042 amend the cursorlineopt setting and allow a 'screenline' value. The value `both` has been aliased to the option `number,line` which is now the new default. To set the value do `:set culopt=screenline` or `:set culopt=screenline,number`. Setting `both,screenline` or `line,screenline` will be an error.

I have been conscious in adjusting the `win_line()` function. It needed quite a bit additional code to handle all different special cases, like when 'list' is set or when a multibyte character does not fit on the screen and gets wrapped and handling are different drawing states. Also, it might cause a small performance penalty, because the caching using `w_last_cursorline` does not work, since we are only highlighting parts of the file.  

BTW: I noticed that the default cursorline highlighting does also underline the `sbr` setting (which is a bit strange, since it does not actually contain to the line content). I almost fixed it, but a test started to fail, so I left it as is (and is only a minor annoyance).

To that end, I added quite a few screen tests, that should make sure that most of the special cases are correctly handled. 

I think most of the special cases are explained in the corresponding commits. I left those commits there and did not squash them to allow for easier review, but I can squash if needed.

chrisbra committed:
refactor parsing of culopt option

Store the actual culopt option value as a flag inside the window
structure, so we can actual OR the different values.

Remove the global p_culopt_values as it is not needed anymore.

chrisbra committed:
Allow screenline as value for cursorlineopt

chrisbra committed:
Initial implementation for screenline highlighting

No configuration yet, just always do screen line highlighting

chrisbra committed:
Force cursorline updating in main_loop and change.c

and also force redraw in change_common if
'culopt' starts with "s"

chrisbra committed:
Correctly reset highlighting after wrapping

If the character continues on the next line, need to correctly reset the
character highlighting so that screenline cursorline is not applied to
the wrapped part of the character. So save and store the previous
char_attr before applying the cursorline highlighting to it.

Once the line wrapps, we need to test if wrapping occurs and a screen
line cursorline highlighting is needed and apply the stored highlighting
rather than the actual char_attr to the next chars.

Also: correctly handle diff mode

Make sure to use the line highlighting and combine it with the
cursorline highlighting attribute.

In addition, correctly highlight parts of previous line
that have wrapped

chrisbra committed:
Only calculate margin columns once

cache values for determining the margins

chrisbra committed:
Apply screenline cursorline to special chars

First of all, we need to add the cursorline screen highlighting to
special items on a line (e.g. like special multibyte characters like
 `<200d>` zero-width joiner or to a tab when list mode is set `^I`.

However this might pose a problem, if this character spans over a line
and is wrapped. Then the cursorline highlighting will be extended on the
next screen line. To prevent this, save and restore the extra_attr
before applying the cursorline highlighting to it and restore it after a
line has been wrapped.

Also add screenline cursorline highlighting in a couple of more places

chrisbra committed:
Re-apply screenline cursorline highlight after wrapped multibyte char

After a partly wrapped multibyte char when the start of the line is
being drawn, make sure to add the cursorline highlighting to that partly
character at the beginning of the line

chrisbra committed:
Document changed values for the cursorlineopt setting

chrisbra committed:
Make use of cursorlineopt flags

replace the test of *wp->w_p_culopt == 'l' etc by making use of the
parsed flags available in wp->w_p_culopt_flags.

chrisbra committed:
cursorline highlighting: add tests

Update option values test for culopt
and add screen tests

dpelle commented on src/option.c:
```diff
@@ -9515,3 +9522,55 @@ get_winbuf_options(int bufopt)
+
+#ifdef FEAT_SYN_HL
+/*
+ * This is called when 'culopt' is change
```
is change → is changed

dpelle commented on runtime/doc/options.txt:
```diff
@@ -2468,17 +2468,21 @@ A jump table for the options with a short description can be found at |Q_op|.
 			CursorLineNr |hl-CursorLineNr|.
-	    "both"	Highlight as both "line" and "number" are set.
+	Special value:
+	"both"		Alias for the values "line,number".
```
"both" is a strange name, now that 'culopt' has more than 2 options. But I understand it's needed for backward compatibility.

dpelle commented on src/screen.c:
```diff
@@ -158,6 +158,9 @@ static void win_redr_custom(win_T *wp, int draw_ruler);
 #endif
+#ifdef FEAT_SYN_HL
+static void margin_columns_win(win_T *wp, int *lmargin, int *rmargin);
+ #endif
```
Indentation is wrong here.

dpelle commented on src/screen.c:
```diff
@@ -5127,6 +5233,11 @@ win_line(
+#ifdef FEAT_SYN_HL
+			if (need_cul_screenline)
+			    extra_attr = hl_combine_attr(extra_attr, HL_ATTR(HLF_CUL));
+ #endif
```
Indentation not aligned with #ifdef

chrisbra committed:
Review

Fix a couple of warnings,
indent #ifdefs appropriately
fix comment

Make bool operator logic in change.c more obvious
Fix wrong initialization in margin_columns_win

brammool commented:
Christian wrote:

> As requested in
> https://github.com/vim/vim/pull/4693#issuecomment-529642042 amend the
> cursorlineopt setting and allow a &#39;screenline&#39; value. The
> value `both` has been aliased to the option `number,line` which is now
> the new default. To set the value do `:set culopt=screenline` or `:set
> culopt=screenline,number`. Setting `both,screenline` or
> `line,screenline` will be an error.
> 
> I have been conscious in adjusting the `win_line()` function. It
> needed quite a bit additional code to handle all different special
> cases, like when &#39;list&#39; is set or when a multibyte character
> does not fit on the screen and gets wrapped and handling are different
> drawing states. Also, it might cause a small performance penalty,
> because the caching using `w_last_cursorline` does not work, since we
> are only highlighting parts of the file.  

I was going to include this, but with a simple test the highlighting is
very wrong.  Try editing src/main.c with:
	hi CursorLine cterm=underline	

With 'cursorlineopt' set to "both" the number column isn't highlighted.
With set to "screenline" lots of highlighting is missing.


-- 
If VIM were a woman, I'd marry her.  Slim, organized, helpful
and beautiful; what's not to like?     --David A. Rogers

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
